File: src/main/java/com/bibireden/data_attributes/DataAttributesClient.java
package com.bibireden.data_attributes;

import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

import com.bibireden.data_attributes.data.AttributeResourceManager;
import com.bibireden.data_attributes.endec.nbt.NbtDeserializer;
import io.netty.util.concurrent.Future;
import io.netty.util.concurrent.GenericFutureListener;
import io.wispforest.endec.format.bytebuf.ByteBufDeserializer;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.networking.v1.ClientLoginNetworking;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.PacketSender;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientLoginNetworkHandler;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.network.PacketByteBuf;

public class DataAttributesClient implements ClientModInitializer {
    // Handles the received login query packet
    private static CompletableFuture<PacketByteBuf> loginQueryReceived(MinecraftClient client, ClientLoginNetworkHandler handler, PacketByteBuf buf, Consumer<GenericFutureListener<? extends Future<? super Void>>> listenerAdder) {
        onPacketReceived(client, buf);
        return CompletableFuture.completedFuture(PacketByteBufs.empty());
    }

    // Handles the received update packet
    private static void updateReceived(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender responseSender) {
        onPacketReceived(client, buf);
    }

    // Common method for handling received packets
    private static void onPacketReceived(MinecraftClient client, PacketByteBuf buf) {
        client.execute(() -> {
            DataAttributes.CLIENT_MANAGER = AttributeResourceManager.ENDEC.decodeFully(ByteBufDeserializer::of, buf);
            DataAttributes.CLIENT_MANAGER.onDataUpdate();
        });
    }

    // Initialization method called when the client starts
    @Override
    public void onInitializeClient() {
        // Register packet handlers for login query and update packets
        ClientLoginNetworking.registerGlobalReceiver(DataAttributes.HANDSHAKE, DataAttributesClient::loginQueryReceived);
        ClientPlayNetworking.registerGlobalReceiver(DataAttributes.RELOAD, DataAttributesClient::updateReceived);
    }
}

File: src/main/kotlin/com/bibireden/data_attributes/DataAttributesClient.kt
package com.bibireden.data_attributes

import com.bibireden.data_attributes.config.AttributeConfigManager
import com.bibireden.data_attributes.config.ConfigUIState
import com.bibireden.data_attributes.networking.NetworkingChannels
import com.bibireden.data_attributes.ui.DataAttributesConfigScreen
import io.wispforest.endec.format.bytebuf.ByteBufDeserializer
import io.wispforest.owo.config.ui.ConfigScreen
import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.networking.v1.ClientLoginNetworking
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.minecraft.client.MinecraftClient
import net.minecraft.network.PacketByteBuf
import java.util.concurrent.CompletableFuture

class DataAttributesClient : ClientModInitializer {
    companion object {
        @JvmField
        val MANAGER = AttributeConfigManager()

        @JvmField
        val UI_STATE = ConfigUIState()

        /** Whenever the client receives a sync packet from the server to update the world-state via. configuration. */
        fun onPacketReceived(client: MinecraftClient, buf: PacketByteBuf) {
            buf.retain()
            client.execute {
                MANAGER.readPacket(AttributeConfigManager.Packet.ENDEC.decodeFully(ByteBufDeserializer::of, buf))
                MANAGER.onDataUpdate()
                buf.release()
            }
        }
    }

    override fun onInitializeClient() {
        ConfigScreen.registerProvider(DataAttributes.MOD_ID) { DataAttributesConfigScreen(DataAttributes.OVERRIDES_CONFIG, DataAttributes.FUNCTIONS_CONFIG, DataAttributes.ENTITY_TYPES_CONFIG, it) }

        ClientLoginNetworking.registerGlobalReceiver(NetworkingChannels.HANDSHAKE) { client, _, buf, _ ->
            onPacketReceived(client, buf)
            CompletableFuture.completedFuture(PacketByteBufs.empty())
        }
        ClientPlayNetworking.registerGlobalReceiver(NetworkingChannels.RELOAD) { client, _, buf, _ ->
            onPacketReceived(client, buf)
        }
    }
}